File: inventory-framework-platform-bukkit/src/main/java/me/devnatan/inventoryframework/context/SlotClickContext.java
package me.devnatan.inventoryframework.context;

import me.devnatan.inventoryframework.RootView;
import me.devnatan.inventoryframework.ViewContainer;
import me.devnatan.inventoryframework.Viewer;
import me.devnatan.inventoryframework.component.Component;
import org.bukkit.entity.Player;
import org.bukkit.event.inventory.ClickType;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.inventory.InventoryType;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.PlayerInventory;
import org.jetbrains.annotations.ApiStatus;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class SlotClickContext extends SlotContext implements IFSlotClickContext {

    private final Viewer whoClicked;
    private final ViewContainer clickedContainer;
    private final Component clickedComponent;
    private final InventoryClickEvent clickOrigin;
    private final boolean combined;
    private boolean cancelled;

    @ApiStatus.Internal
    public SlotClickContext(
            int slot,
            @NotNull IFRenderContext parent,
            @NotNull Viewer whoClicked,
            @NotNull ViewContainer clickedContainer,
            @Nullable Component clickedComponent,
            @NotNull InventoryClickEvent clickOrigin,
            boolean combined) {
        super(slot, parent);
        this.whoClicked = whoClicked;
        this.clickedContainer = clickedContainer;
        this.clickedComponent = clickedComponent;
        this.clickOrigin = clickOrigin;
        this.combined = combined;
    }

    /**
     * The player who clicked on the slot.
     */
    public final @NotNull Player getPlayer() {
        return (Player) clickOrigin.getWhoClicked();
    }

    /**
     * The event that triggered this context.
     * <p>
     * This is an internal inventory-framework API that should not be used from outside of
     * this library. No compatibility guarantees are provided.
     */
    @NotNull
    public InventoryClickEvent getClickOrigin() {
        return clickOrigin;
    }

    /**
     * The item that was clicked.
     */
    @Override
    public final ItemStack getItem() {
        return clickOrigin.getCurrentItem();
    }

    @Override
    public final Component getComponent() {
        return clickedComponent;
    }

    @Override
    public final @NotNull ViewContainer getClickedContainer() {
        return clickedContainer;
    }

    @Override
    public final boolean isCancelled() {
        return cancelled;
    }

    @Override
    public final void setCancelled(boolean cancelled) {
        this.cancelled = cancelled;
        getClickOrigin().setCancelled(cancelled);
    }

    @Override
    public final Object getPlatformEvent() {
        return clickOrigin;
    }

    @Override
    public final int getClickedSlot() {
        return clickOrigin.getRawSlot();
    }

    @Override
    public final boolean isLeftClick() {
        return getClickOrigin().isLeftClick();
    }

    @Override
    public final boolean isRightClick() {
        return getClickOrigin().isRightClick();
    }

    @Override
    public final boolean isMiddleClick() {
        return getClickOrigin().getClick() == ClickType.MIDDLE;
    }

    @Override
    public final boolean isShiftClick() {
        return getClickOrigin().isShiftClick();
    }

    @Override
    public final boolean isKeyboardClick() {
        return getClickOrigin().getClick().isKeyboardClick();
    }

    @Override
    public final boolean isOutsideClick() {
        return getClickOrigin().getSlotType() == InventoryType.SlotType.OUTSIDE;
    }

    @Override
    public final String getClickIdentifier() {
        return getClickOrigin().getClick().name();
    }

    @Override
    public final boolean isOnEntityContainer() {
        return getClickOrigin().getClickedInventory() instanceof PlayerInventory;
    }

    @Override
    public final Viewer getViewer() {
        return whoClicked;
    }

    @Override
    public final void closeForPlayer() {
        getParent().closeForPlayer();
    }

    @Override
    public final void openForPlayer(@NotNull Class<? extends RootView> other) {
        getParent().openForPlayer(other);
    }

    @Override
    public final void openForPlayer(@NotNull Class<? extends RootView> other, Object initialData) {
        getParent().openForPlayer(other, initialData);
    }

    @Override
    public final void updateTitleForPlayer(@NotNull String title) {
        getParent().updateTitleForPlayer(title);
    }

    @Override
    public final void updateTitleForPlayer(@NotNull Object title) {
        getParent().updateTitleForPlayer(title);
    }

    @Override
    public final void resetTitleForPlayer() {
        getParent().resetTitleForPlayer();
    }

    @Override
    public final boolean isCombined() {
        return combined;
    }
}

File: inventory-framework-platform-minestom/src/main/kotlin/me/devnatan/inventoryframework/context/SlotClickContext.kt
package me.devnatan.inventoryframework.context

import me.devnatan.inventoryframework.RootView
import me.devnatan.inventoryframework.ViewContainer
import me.devnatan.inventoryframework.Viewer
import me.devnatan.inventoryframework.component.Component
import net.minestom.server.entity.Player
import net.minestom.server.event.inventory.InventoryPreClickEvent
import net.minestom.server.inventory.PlayerInventory
import net.minestom.server.inventory.click.Click
import net.minestom.server.item.ItemStack
import org.jetbrains.annotations.ApiStatus

class SlotClickContext
    @ApiStatus.Internal
    constructor(
        slot: Int,
        parent: IFRenderContext,
        private val whoClicked: Viewer,
        private val clickedContainer: ViewContainer,
        private val clickedComponent: Component?,
        val clickOrigin: InventoryPreClickEvent,
        private val combined: Boolean,
    ) : SlotContext(slot, parent),
        IFSlotClickContext {
        private var cancelled = false

        override val player: Player
            /** The player who clicked on the slot. */
            get() = clickOrigin.player

        override val item: ItemStack
            /** The item that was clicked. */
            get() = clickOrigin.clickedItem

        override fun getComponent(): Component? = clickedComponent

        override fun getClickedContainer(): ViewContainer = clickedContainer

        override fun isCancelled(): Boolean = cancelled

        override fun setCancelled(cancelled: Boolean) {
            this.cancelled = cancelled
            clickOrigin.isCancelled = cancelled
        }

        override fun getPlatformEvent(): Any = clickOrigin

        override fun getClickedSlot(): Int = clickOrigin.slot

        override fun isLeftClick(): Boolean = clickOrigin.click is Click.Left

        override fun isRightClick(): Boolean = clickOrigin.click is Click.Right

        override fun isMiddleClick(): Boolean = false

        override fun isShiftClick(): Boolean {
            val clickType = clickOrigin.click
            return clickType is Click.LeftShift || clickType is Click.RightShift
        }

        override fun isKeyboardClick(): Boolean = clickOrigin.click is Click.HotbarSwap

        override fun isOutsideClick(): Boolean = clickOrigin.slot < 0

        override fun getClickIdentifier(): String =
            clickOrigin.click.javaClass.simpleName
                .lowercase()

        override fun isOnEntityContainer(): Boolean = clickOrigin.inventory is PlayerInventory

        override fun getViewer(): Viewer = whoClicked

        override fun closeForPlayer() {
            parent.closeForPlayer()
        }

        override fun openForPlayer(other: Class<out RootView?>) {
            parent.openForPlayer(other)
        }

        override fun openForPlayer(
            other: Class<out RootView?>,
            initialData: Any,
        ) {
            parent.openForPlayer(other, initialData)
        }

        override fun updateTitleForPlayer(title: String) {
            parent.updateTitleForPlayer(title)
        }

        override fun resetTitleForPlayer() {
            parent.resetTitleForPlayer()
        }

        override fun isCombined(): Boolean = combined
    }
